Sources/SwiftKafka/Utilities/KafkaStatisticsJsonModel.swift

```diff
@@ -0,0 +1,177 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the swift-kafka-gsoc open source project
+//
+// Copyright (c) 2022 Apple Inc. and the swift-kafka-gsoc project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of swift-kafka-gsoc project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+// This file was generated from JSON Schema using quicktype, do not modify it directly.
+// To parse the JSON, add this file to your project and do:
+//
+//   let statistics = try? newJSONDecoder().decode(KafkaStatisticsJsonModel.self, from: jsonData)
+
+// MARK: - Statistics
+
+public struct KafkaStatisticsJson: Hashable, Codable {
```

Comment:
Thank you for your comment!

I think that `swift-metrics` is a good thing and should be used to deliver statistics and that is the most important thing.

However, from my perspective, there could be other scenarios when statistics can be used by an end-user application directly for adjusting its behaviour. 
There are some examples: 
1. measure lag to the final offset to estimate [when/are/time for] all records received 
2. work with latency metrics to provide approximate latency for end user
3. understand the number messages in flight to adjust internal throttling/coalescing/backpressure

I believe there could be other various scenarios when it is required by end-user application. However..

Next thing that comes to mind is about delivery of this statistics to `swift-metrics`: I am not sure if there is an automatic way to deliver the whole json to swift-metics directly...
Therefore, I am a little bit confused if there is a suggestion to use some other model instead of proposed. 
Should we?

Reply:
Hello @blindspotbounty ,

I don't think providing metrics to an end-user application is a priority. Users should just tweak their configurations to work well and should not have some statistics-dependent magic running in their applications IMO.

Unfortunately, there is no direct mapping from `json` to `swift-metrics`. However, here is how I would go about implementing this:

1. Have a new type called `KafkaMetricsReporting` that you pass as an _optional_ argument as part of `KafkaConsumerConfiguration`/`KafkaProducerConfiguration`

This could look something like this:

```swift
public struct KafkaMetricsReporting {
    public let myCounter: Metrics.Counter?
    public let myGauge: Metrics.Gauge?
    public let myTimer: Metrics.Timer?
}
```

```swift
public struct KafkaConsumerConfig {
    public var metrics: KafkaMetricsReporting?
}
```

This can be similar to the `KafkaStatistics` type you already have. However, you'd have to find suitable [metric types](https://github.com/apple/swift-metrics#metric-types) for each statistics property.

2. If the `Kafka[Consumer|Producer]Configuration` is non-`nil` and not all of its properties are `nil`, enable for statistics and update the metrics accordingly when a new statistics event is received

3. When you have received the event, update all metrics in you `KafkaMetricsReporting` that are non-`nil`

Reply:
That sounds nice and should be pretty much convenient.
However, there could be a number of topics and partitions in them. Currently, I see just several solutions that might suit here:
1. Provide metrics factory that would make counters with labels topic=abc partition=xx
2. Calculate averages/totals for all topics/partitions for specific counters (probably the best)
3. Provide different gauges/counters in a dictionary to configuration
4. Don't cover array fields so far
 
Maybe there is some better solution, let me know if you have something in mind, please.

Reply:
Hey @blindspotbounty , I haven’t really thought about how we should handle these array based metrics for topic / partitions etc. however I am inclined to not aggregate them to averages but rather expose topic / partition … -specific information.

Metrics has evolved into quite a big feature now so I would suggest not covering all metrics at once but rather start by getting the general structure right and listening to a hand full of metrics.

You can then just add the remaining metrics in a follow-up pull request once we get the basics right :smile:

I believe starting with some of the [Top Level Metrics](https://github.com/confluentinc/librdkafka/blob/master/STATISTICS.md#top-level) is a good start